src/plugins/tonegen/plugin.c

```diff
@@ -23,74 +26,177 @@
+    int     standard;
+    int     sample_rate;
+    int     statistics;
+    int     buflen;
+    int     minreq;
+    char    *dtmf_tags;
+    char    *notif_tags;
+    char    *ind_tags;
+    int     dtmf_volume;
+    int     notif_volume;
+    int     ind_volume;
+    int     legacy_api;
+};
+
+struct userdata {
+    NPlugin *plugin;
+    struct properties properties;
+    struct tonegend tonegend;
+};
+
+struct options_parse;
+typedef int (*prop_value_parser)(const NValue *val, struct options_parse *opt);
+
+struct options_parse {
+    const char *arg_name;
+    prop_value_parser arg_parser;
+    int *arg_int;
+    char **arg_str;
+};
+
+static struct userdata u;
+
+static int prop_8khz_parser(const NValue *val, struct options_parse *opt);
+static int prop_standard_parser(const NValue *val, struct options_parse *opt);
+static int prop_string_parser(const NValue *val, struct options_parse *opt);
+static int prop_int_parser(const NValue *val, struct options_parse *opt);
+static int prop_bool_parser(const NValue *val, struct options_parse *opt);
+
+static struct options_parse options[] = {
+    { "8kHz"            , prop_8khz_parser      , &u.properties.sample_rate, NULL },
+    { "standard"        , prop_standard_parser  , &u.properties.standard, NULL    },
+    { "buflen"          , prop_int_parser       , &u.properties.buflen, NULL      },
+    { "minreq"          , prop_int_parser       , &u.properties.minreq, NULL      },
+    { "statistics"      , prop_bool_parser      , &u.properties.statistics, NULL  },
+    { "tag-dtmf"        , prop_string_parser    , NULL, &u.properties.dtmf_tags   },
+    { "tag-indicator"   , prop_string_parser    , NULL, &u.properties.ind_tags    },
+    { "tag-notif"       , prop_string_parser    , NULL, &u.properties.notif_tags  },
+    { "volume-dtmf"     , prop_int_parser       , &u.properties.dtmf_volume, NULL },
+    { "volume-indicator", prop_int_parser       , &u.properties.notif_volume, NULL},
+    { "volume-notif"    , prop_int_parser       , &u.properties.ind_volume, NULL  },
+    { "legacy-dbus-api" , prop_bool_parser      , &u.properties.legacy_api, NULL  },
+    { NULL              , NULL                  , NULL, NULL                      }
+};
+
+/* OPTION PARSING */
 
-static DBusConnection *system_bus = NULL;
+static int
+prop_8khz_parser (const NValue *val, struct options_parse *opt)
+{
+    if (prop_bool_parser (val, opt)) {
+        if (*opt->arg_int == 1)
+            *opt->arg_int = 8000;
+        return 1;
+    } else
+        return 0;
+}
 
-static gboolean
-call_dbus_method (DBusConnection *bus, DBusMessage *msg)
+static int
+prop_bool_parser (const NValue *val, struct options_parse *opt)
 {
-    if (!dbus_connection_send (bus, msg, 0)) {
-        N_WARNING (LOG_CAT "failed to send DBus message %s",
-            dbus_message_get_member (msg), dbus_message_get_interface (msg));
+    const char *str;
+
+    str = n_value_get_string (val);
+    if (g_strcmp0 (str, "true") == 0 ||
```

Comment:
Or just `g_strcasecmp(str ?: "", "true")` to cover all possible lowercase/uppercase combinations of "true"?